src/undo_log/undo_log.c

```diff
@@ -0,0 +1,290 @@
+}
+
+// undo edge creation
+void UndoLog_CreateEdge
+(
+	UndoLog *log,
+	Edge edge             // edge created
+) {
+	ASSERT(op != NULL && *op != NULL);
+
+	UndoOp op;
+
+	op.type        = UL_CREATE_EDGE;
+	op.create_op.e = edge;
+
+	*log = array_append(*log, op);
+}
+
+// undo node deletion
+void UndoLog_DeleteNode
+(
+	UndoLog *log,       // undo log
+	Node node,          // node deleted
+	LabelID *labelIDs,  // node labels
+	uint label_count    // node label count
+) {
+	ASSERT(log != NULL && *log != NULL);
+
+	UndoOp op;
+
+	op.type                   =  UL_DELETE_NODE;
+	op.delete_op.n            =  node;
+	op.delete_op.labels       =  labelIDs;
+	op.delete_op.label_count  =  label_count;
+
+	*log = array_append(*log, op);
+}
+
+// undo edge deletion
+void UndoLog_DeleteEdge
+(
+	UndoLog *log,  // undo log
+	Edge edge      // edge deleted
+) {
+	ASSERT(op != NULL && *op != NULL);
+
+	UndoOp op;
+
+	op.type        = UL_DELETE_EDGE;
+	op.delete_op.e = edge;
+
+	*log = array_append(*log, op);
+}
+
+// undo entity update
+void UndoLog_UpdateEntity
+(
+	UndoLog *log,                // undo log
+	GraphEntity *ge,             // updated entity
+	Attribute_ID attr_id,        // updated attribute ID
+	SIValue orig_value,          // attribute original value
+	GraphEntityType entity_type  // entity type
+) {
+	ASSERT(log != NULL && *log != NULL);
+	ASSERT(ge != NULL);
+	ASSERT(attr_id != ATTRIBUTE_ID_NONE && attr_id != ATTRIBUTE_ID_ALL);
+
+	UndoOp op;
+
+	op.type                  = UL_UPDATE;
+	op.update_op.ge          = ge;
+	op.update_op.attr_id     = attr_id;
+	op.update_op.orig_value  = orig_value;
+	op.update_op.entity_type = entity_type;
+
+	*log = array_append(*log, op);
+}
+//------------------------------------------------------------------------------
+// rollback
+//------------------------------------------------------------------------------
+
+void UndoLog_Rollback
+(
+	UndoLog log
+) {
+	QueryCtx *ctx = QueryCtx_GetQueryCtx();
+	ASSERT(ctx != NULL);
+	ASSERT(log != NULL);
+
+	uint64_t count = array_len(undo_list);
+
+	// apply undo operations is reverse order for rollback correctness
+	// Find sequences of operation of the same type and rollback them as a bulk
+	for (uint i = count-1; i >= 0 ; i--) {
+		UndoOpType cur_type = undo_list[i].type;
+		uint seq_end = i;
+		while(seq_end - 1 >= 0 && undo_list[seq_end - 1].type == cur_type) {
+			seq_end--;
+		}
+
+		switch(cur_type) {
+			case UL_UPDATE:
+				_UndoLog_Rollback_Update_Entity(ctx, i, seq_end);
+				break;
+			case UL_CREATE_NODE:
+				_UndoLog_Rollback_Create_Node(ctx, i, seq_end);
+				break;
+			case UL_CREATE_EDGE:
+				_UndoLog_Rollback_Create_Edge(ctx, i, seq_end);
+				break;
+			case UL_DELETE_NODE:
+				_UndoLog_Rollback_Delete_Node(ctx, i, seq_end);
+				break;
+			case UL_DELETE_EDGE:
+				_UndoLog_Rollback_Delete_Edge(ctx, i, seq_end);
+				break;
+			default:
+				ASSERT(false);           
+		}
+
+		i = seq_end;
+	}
+
+	array_clear(undo_list);
```

Comment:
in-case the undo log is reused from one query to another please note that `array_clear` doesn't shrink the array, so in-case we had a large undo-log we wouldn't return the memory consumed by it.